kmp-library/src/commonMain/kotlin/com/novoda/aws/v4/signer/CanonicalRequest.kt

```diff
@@ -0,0 +1,145 @@
+/*
+  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+  the License. You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+  specific language governing permissions and limitations under the License.
+
+  Copyright 2016 the original author or authors.
+ */
+package com.novoda.aws.v4.signer
+
+import com.novoda.aws.v4.signer.encoding.URLEncoding
+
+private const val S3_SERVICE = "s3"
+private const val QUERY_PARAMETER_SEPARATOR = '&'
+private const val QUERY_PARAMETER_VALUE_SEPARATOR = '='
+
+/**
+ * @author Richard Lucas
+ */
+class CanonicalRequest(
+        private val service: String,
+        private val httpRequest: HttpRequest,
+        val headers: CanonicalHeaders,
+        private val contentSha256: String
+) {
+
+    fun get(): String {
+        return httpRequest.method +
+                "\n" + normalizePath(httpRequest.path) +
+                "\n" + normalizeQuery(httpRequest.query) +
+                "\n" + headers.canonicalizedHeaders +
+                "\n" + headers.names +
+                "\n" + contentSha256
+    }
+
+    override fun toString(): String {
+        return get()
+    }
+
+    private fun normalizePath(path: String?): String {
+        if (path.isNullOrEmpty()) {
+            return "/"
+        }
+        // Encode characters as mandated by AWS
+        val encoded = URLEncoding.encodePath(path)
+        if (S3_SERVICE == service) {
+            /*
+             * S3 requests should not be normalized.
+             * See http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html#canonical-request
+             */
+            return encoded
+        }
+        return PathUtil.normalize(encoded)
+    }
+}
+
+private fun normalizeQuery(rawQuery: String?): String {
+    if (rawQuery.isNullOrEmpty()) {
+        return ""
+    }
+
+    /*
+     * Sort query parameters. Simply sort lexicographically by character
+     * code, which is equivalent to comparing code points (as mandated by
+     * AWS)
+     */
+    return rawQuery
+            .extractQueryParameters()
+            .sortedBy(Pair<String, String?>::first)
+            .fold(StringBuilder()) { builder, parameter ->
+                val name = parameter.first
+                val value = parameter.second ?: "" // if no value use an empty string as per the spec
```

Comment:
I don't really feel that comfortable changing this class much because there are no tests for this bit of code, I like the deconstruction in lambda though that's safe 👍 .

Regarding the constants, I extracted them because they are used in other parts of the class

Reply:
Yeah, sounds good as we don't have tests around it